1) 데이터 준비와 전처리
---
Movielens 데이터는 rating.dat 안에 이미 인덱싱까지 완료된 사용자-영화-평점 데이터가 깔끔하게 정리되어 있습니다.

루브릭

아래의 기준을 바탕으로 프로젝트를 평가합니다.


|평가문항|	상세기준|
|:------:|:--------:|
|1. CSR matrix가 정상적으로 만들어졌다.| 사용자와 아이템 개수를 바탕으로 정확한 사이즈로 만들었다.|
|2. MF 모델이 정상적으로 훈련되어 그럴듯한 추천이 이루어졌다.| 사용자와 아이템 벡터 내적수치가 의미있게 형성되었다.|
|3. 비슷한 영화 찾기와 유저에게 추천하기의 과정이 정상적으로 진행되었다. |MF모델이 예측한 유저 선호도 및 아이템간 유사도, 기여도가 의미있게 측정되었다.|

In [387]:
import os
import pandas as pd

rating_file_path=os.getenv('HOME') + '/aiffel/recommendata_iu/data/ml-1m/ratings.dat'
ratings_cols = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_csv(rating_file_path, sep='::', names=ratings_cols, engine='python', encoding = "ISO-8859-1")
orginal_data_size = len(ratings)
ratings.head()

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

In [388]:
# 3점 이상만 남깁니다.
ratings = ratings[ratings['rating']>=3]
filtered_data_size = len(ratings)

print(f'orginal_data_size: {orginal_data_size}, filtered_data_size: {filtered_data_size}')
print(f'Ratio of Remaining Data is {filtered_data_size / orginal_data_size:.2%}')

orginal_data_size: 1000209, filtered_data_size: 836478
Ratio of Remaining Data is 83.63%


In [389]:
# rating 컬럼의 이름을 count로 바꿉니다.
ratings.rename(columns={'rating':'count'}, inplace=True)

In [390]:
ratings['count']

0          5
1          3
2          3
3          4
4          5
          ..
1000203    3
1000205    5
1000206    5
1000207    4
1000208    4
Name: count, Length: 836478, dtype: int64

In [391]:
# 영화 제목을 보기 위해 메타 데이터를 읽어옵니다.
movie_file_path=os.getenv('HOME') + '/aiffel/recommendata_iu/data/ml-1m/movies.dat'
cols = ['movie_id', 'title', 'genre'] 
movies = pd.read_csv(movie_file_path, sep='::', names=cols, engine='python', encoding='ISO-8859-1')
movies.head()

movie_id                               title                         genre
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

2) 분석해 봅시다.

In [392]:
# movie_id는 3952까지 존재
movies.tail()

movie_id                       title           genre
3878      3948     Meet the Parents (2000)          Comedy
3879      3949  Requiem for a Dream (2000)           Drama
3880      3950            Tigerland (2000)           Drama
3881      3951     Two Family House (2000)           Drama
3882      3952       Contender, The (2000)  Drama|Thriller

In [393]:
# 1) ratings에 있는 유니크한 영화 개수
ratings['movie_id'].nunique()

3628

In [394]:
# 2) rating에 있는 유니크한 사용자 수
ratings['user_id'].nunique()

6039

In [395]:
# 3) 가장 인기 있는 영화 30개(인기순)
ratings.groupby('movie_id')['user_id'].count().sort_values(ascending=False)

movie_id
2858    3211
260     2910
1196    2885
1210    2716
2028    2561
        ... 
1553       1
1548       1
2486       1
138        1
3876       1
Name: user_id, Length: 3628, dtype: int64

In [396]:
ratings

user_id  movie_id  count  timestamp
0              1      1193      5  978300760
1              1       661      3  978302109
2              1       914      3  978301968
3              1      3408      4  978300275
4              1      2355      5  978824291
...          ...       ...    ...        ...
1000203     6040      1090      3  956715518
1000205     6040      1094      5  956704887
1000206     6040       562      5  956704746
1000207     6040      1096      4  956715648
1000208     6040      1097      4  956715569

[836478 rows x 4 columns]

In [398]:
ratings.tail()

user_id  movie_id  count
1000203     6040      1090      3
1000205     6040      1094      5
1000206     6040       562      5
1000207     6040      1096      4
1000208     6040      1097      4

In [399]:
ratings = ratings.merge(movies, how='left', on='movie_id')

In [400]:
movie = 'E.T. the Extra-Terrestrial (1982)'

In [401]:
ratings[ratings['title'].apply(lambda x : x.split('(')[0].rstrip() in movie)]

user_id  movie_id  count                              title  \
26            1      1097      4  E.T. the Extra-Terrestrial (1982)   
225           4      1097      4  E.T. the Extra-Terrestrial (1982)   
1076         10      1097      5  E.T. the Extra-Terrestrial (1982)   
1309         13      1097      5  E.T. the Extra-Terrestrial (1982)   
1712         17      1097      3  E.T. the Extra-Terrestrial (1982)   
...         ...       ...    ...                                ...   
835166     6035      1097      4  E.T. the Extra-Terrestrial (1982)   
835873     6036      1097      4  E.T. the Extra-Terrestrial (1982)   
836064     6037      1097      5  E.T. the Extra-Terrestrial (1982)   
836201     6039      1097      4  E.T. the Extra-Terrestrial (1982)   
836477     6040      1097      4  E.T. the Extra-Terrestrial (1982)   

                                  genre  
26      Children's|Drama|Fantasy|Sci-Fi  
225     Children's|Drama|Fantasy|Sci-Fi  
1076    Children's|Drama|Fantasy|Sci-Fi  
1309    Children's|Drama|Fantasy|Sci-Fi  
1712    Children's|Drama|Fantasy|Sci-Fi  
...                                 ...  
835166  Children's|Drama|Fantasy|Sci-Fi  
835873  Children's|Drama|Fantasy|Sci-Fi  
836064  Children's|Drama|Fantasy|Sci-Fi  
836201  Children's|Drama|Fantasy|Sci-Fi  
836477  Children's|Drama|Fantasy|Sci-Fi  

[2102 rows x 5 columns]

In [402]:
# 연도는 1919년 부터 2000년까지 존재
ratings['title'].apply(lambda x : int(x.split('(')[-1][:4])).sort_values()

371730    1919
395335    1919
835412    1919
587578    1919
665294    1919
          ... 
413662    2000
77813     2000
309771    2000
309769    2000
762702    2000
Name: title, Length: 836478, dtype: int64

In [403]:
genres = list(movies['genre'].value_counts().keys())

In [404]:
def unique_genre(genres):
    uniq_genre = [] # unique 장르를 담을 list
    for genre in genres:
        uniq_genre.extend(genre.split('|'))
    return list(set(uniq_genre))

In [405]:
genres = unique_genre(genres)

In [406]:
# G enres are pipe-separated and are selected from the following genres:
genres.sort()
genres

['Action',
 'Adventure',
 'Animation',
 "Children's",
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Fantasy',
 'Film-Noir',
 'Horror',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Thriller',
 'War',
 'Western']

3) 내가 선호하는 영화를 5가지 골라서 rating에 추가해 줍시다.

- user_id : 기존 id가 6040까지 있으므로 내 id는 6041으로 설정한다.
- movie_id : 내가 선택할 영화의 id가 리스트에 존재한다면 해당 id로 설정하고
그렇지 않다면 3952 이후로 설정 
- count : 내가 좋아하는 영화이므로 5점을 부여
- title : title 형식에 맞게 '제목 (연도)'로 전처리
- genre : 장르가 여러개 존재할 경우 |로 연결 

In [407]:
# 2000년 이전 영화 중에서 내가 좋아하는 영화 5가지를 선택한다.

# 라이언 일병 구하기 Saving Private Ryan, 1998 War, Action, Drama
# 인생은 아름다워 Life Is Beautiful, 1997, Drama, Comedy
# 쇼생크 탈출 The Shawshank Redemption, 1994, Drama
# 타이타닉 Titanic, 1997, Romance, Drama
# 매트릭스 The Matrix, 1999 Sci-Fi, Action

# 영화 하나 하나 검색하며 title이 일치하는지 여부를 확인한다.

favorite_movies = ['Saving Private Ryan (1998)',
                   'Life Is Beautiful (1997)',
                   'The Shawshank Redemption (1994)',
                   'Titanic (1997)',
                   'The Matrix (1999)']

In [408]:
# 라이언 일병 구하기
movie_name = 'Saving Private Ryan (1998)'
ratings[ratings['title'].apply(lambda x : x == movie_name)]

user_id  movie_id  count                       title             genre
48            1      2028      5  Saving Private Ryan (1998)  Action|Drama|War
102           2      2028      4  Saving Private Ryan (1998)  Action|Drama|War
221           4      2028      5  Saving Private Ryan (1998)  Action|Drama|War
473           7      2028      5  Saving Private Ryan (1998)  Action|Drama|War
571           8      2028      5  Saving Private Ryan (1998)  Action|Drama|War
...         ...       ...    ...                         ...               ...
834644     6027      2028      5  Saving Private Ryan (1998)  Action|Drama|War
835008     6033      2028      5  Saving Private Ryan (1998)  Action|Drama|War
835862     6036      2028      5  Saving Private Ryan (1998)  Action|Drama|War
836059     6037      2028      4  Saving Private Ryan (1998)  Action|Drama|War
836470     6040      2028      5  Saving Private Ryan (1998)  Action|Drama|War

[2561 rows x 5 columns]

In [409]:
# 인생은 아름다워
# Life Is Beautiful (1997)으로는 인덱싱이 되지 않는다'
# 제목을 'Life Is Beautiful (La Vita è bella) (1997)'로 수정한다.

movie_name = 'Life Is Beautiful (1997)'
favorite_movies[1] = 'Life Is Beautiful (La Vita è bella) (1997)'
ratings[ratings['title'].apply(lambda x : x[:4] == 'Life')]

user_id  movie_id  count                                       title  \
595           8      2324      3  Life Is Beautiful (La Vita è bella) (1997)   
698           9      2324      5  Life Is Beautiful (La Vita è bella) (1997)   
885          10      2324      5  Life Is Beautiful (La Vita è bella) (1997)   
1673         17      2377      5                            Lifeforce (1985)   
2078         19      2324      5  Life Is Beautiful (La Vita è bella) (1997)   
...         ...       ...    ...                                         ...   
835133     6035      2324      5  Life Is Beautiful (La Vita è bella) (1997)   
835728     6036      2324      4  Life Is Beautiful (La Vita è bella) (1997)   
835756     6036      2377      3                            Lifeforce (1985)   
836038     6037      2324      4  Life Is Beautiful (La Vita è bella) (1997)   
836423     6040      2324      3  Life Is Beautiful (La Vita è bella) (1997)   

                genre  
595      Comedy|Drama  
698      Comedy|Drama  
885      Comedy|Drama  
1673    Horror|Sci-Fi  
2078     Comedy|Drama  
...               ...  
835133   Comedy|Drama  
835728   Comedy|Drama  
835756  Horror|Sci-Fi  
836038   Comedy|Drama  
836423   Comedy|Drama  

[1682 rows x 5 columns]

In [410]:
# 쇼생크 탈출
# 쇼생크 탈출 제목도 Shawshank Redemption, The로 되어있기 떄문에 제목을 수정한다.
movie_name = 'Shawshank Redemption (1994)'
movies[2] = 'Shawshank Redemption, The (1994)'
movies[movies['title'].apply(lambda x : (x[:4] == 'Shaw'))] 

movie_id                             title  genre  \
315       318  Shawshank Redemption, The (1994)  Drama   

                                    2  
315  Shawshank Redemption, The (1994)

In [411]:
# 타이타닉 
movie_name = 'Titanic (1997)'
ratings[ratings['title'].apply(lambda x : x == movie_name)]

user_id  movie_id  count           title          genre
27            1      1721      4  Titanic (1997)  Drama|Romance
557           8      1721      5  Titanic (1997)  Drama|Romance
672           9      1721      5  Titanic (1997)  Drama|Romance
1027         10      1721      3  Titanic (1997)  Drama|Romance
1934         18      1721      4  Titanic (1997)  Drama|Romance
...         ...       ...    ...             ...            ...
833544     6016      1721      3  Titanic (1997)  Drama|Romance
834214     6023      1721      5  Titanic (1997)  Drama|Romance
834500     6025      1721      4  Titanic (1997)  Drama|Romance
834642     6027      1721      4  Titanic (1997)  Drama|Romance
835757     6036      1721      4  Titanic (1997)  Drama|Romance

[1270 rows x 5 columns]

In [412]:
# 매트릭스
# 매트릭스는 Matrix, The (1999)로 수정한다.
movie_name = 'The Matrix (1999)'
favorite_movies[4] = 'Matrix, The (1999)' 
ratings[ratings['title'].apply(lambda x : x[:6] == 'Matrix')]

user_id  movie_id  count               title                   genre
127           2      2571      4  Matrix, The (1999)  Action|Sci-Fi|Thriller
286           5      2571      5  Matrix, The (1999)  Action|Sci-Fi|Thriller
459           7      2571      5  Matrix, The (1999)  Action|Sci-Fi|Thriller
517           8      2571      5  Matrix, The (1999)  Action|Sci-Fi|Thriller
640           9      2571      5  Matrix, The (1999)  Action|Sci-Fi|Thriller
...         ...       ...    ...                 ...                     ...
834757     6030      2571      5  Matrix, The (1999)  Action|Sci-Fi|Thriller
834829     6031      2571      5  Matrix, The (1999)  Action|Sci-Fi|Thriller
835072     6035      2571      5  Matrix, The (1999)  Action|Sci-Fi|Thriller
835456     6036      2571      3  Matrix, The (1999)  Action|Sci-Fi|Thriller
836289     6040      2571      4  Matrix, The (1999)  Action|Sci-Fi|Thriller

[2434 rows x 5 columns]

In [413]:
favorite_movies

['Saving Private Ryan (1998)',
 'Life Is Beautiful (La Vita è bella) (1997)',
 'The Shawshank Redemption (1994)',
 'Titanic (1997)',
 'Matrix, The (1999)']

In [414]:
my_movies = {
    'user_id' : [6041] * 5,
    'movie_id' : [2028, 2324, 318, 1721, 2571],
    'count' : [5]*5,
    'title' : favorite_movies,
    'genre': ['Action|Drama|War',
             'Comedy|Drama',
             'Drama',
             'Drama|Romance',
             'Action|Sci-Fi|Thriller']
}

In [415]:
my_movies_df = pd.DataFrame(my_movies)

In [416]:
ratings = ratings.append(my_movies_df, ignore_index=True)

In [421]:
ratings.tail(10)

user_id  movie_id  count                                       title  \
836473     6040      1090      3                              Platoon (1986)   
836474     6040      1094      5                     Crying Game, The (1992)   
836475     6040       562      5             Welcome to the Dollhouse (1995)   
836476     6040      1096      4                      Sophie's Choice (1982)   
836477     6040      1097      4           E.T. the Extra-Terrestrial (1982)   
836478     6041      2028      5                  Saving Private Ryan (1998)   
836479     6041      2324      5  Life Is Beautiful (La Vita è bella) (1997)   
836480     6041       318      5             The Shawshank Redemption (1994)   
836481     6041      1721      5                              Titanic (1997)   
836482     6041      2571      5                          Matrix, The (1999)   

                                  genre  
836473                        Drama|War  
836474                Drama|Romance|War  
836475                     Comedy|Drama  
836476                            Drama  
836477  Children's|Drama|Fantasy|Sci-Fi  
836478                 Action|Drama|War  
836479                     Comedy|Drama  
836480                            Drama  
836481                    Drama|Romance  
836482           Action|Sci-Fi|Thriller

In [422]:
ratings.drop(['timestamp', 'genre'], axis=1, inplace=True)

In [423]:
ratings

user_id  movie_id  count                                       title
0             1      1193      5      One Flew Over the Cuckoo's Nest (1975)
1             1       661      3            James and the Giant Peach (1996)
2             1       914      3                         My Fair Lady (1964)
3             1      3408      4                      Erin Brockovich (2000)
4             1      2355      5                        Bug's Life, A (1998)
...         ...       ...    ...                                         ...
836478     6041      2028      5                  Saving Private Ryan (1998)
836479     6041      2324      5  Life Is Beautiful (La Vita è bella) (1997)
836480     6041       318      5             The Shawshank Redemption (1994)
836481     6041      1721      5                              Titanic (1997)
836482     6041      2571      5                          Matrix, The (1999)

[836483 rows x 4 columns]

In [424]:
# 고유한 유저와 영화의 개수
user_unique = ratings['user_id'].unique()
movie_unique = ratings['title'].unique()

# 고유한 값의 개수만큼 유저 id와 영화 제목을 indexing
user_to_idx = {v:k for k,v in enumerate(user_unique)}
movie_to_idx = {v:k for k,v in enumerate(movie_unique)}

In [425]:
# 데이터 컬럼 내 값을 indexing된 값으로 교체

# user_to_idx.get을 통해 user_id 컬럼의 모든 값을 인덱싱한 Series를 구해 봅시다. 
# 혹시 정상적으로 인덱싱되지 않은 row가 있다면 인덱스가 NaN이 될 테니 dropna()로 제거합니다. 
temp_user_data = ratings['user_id'].map(user_to_idx.get).dropna()
if len(temp_user_data) == len(ratings):   # 모든 row가 정상적으로 인덱싱되었다면
    print('user_id column indexing OK!!')
    ratings['user_id'] = temp_user_data   # data['user_id']을 인덱싱된 Series로 교체해 줍니다. 
else:
    print('user_id column indexing Fail!!')

# movie_to_idx을 통해 title 컬럼도 동일한 방식으로 인덱싱해 줍니다. 
temp_movie_data = ratings['title'].map(movie_to_idx.get).dropna()
if len(temp_movie_data) == len(ratings):
    print('title column indexing OK!!')
    ratings['title'] = temp_movie_data
else:
    print('title column indexing Fail!!')

ratings

user_id column indexing OK!!
title column indexing OK!!


user_id  movie_id  count  title
0             0      1193      5      0
1             0       661      3      1
2             0       914      3      2
3             0      3408      4      3
4             0      2355      5      4
...         ...       ...    ...    ...
836478     6039      2028      5     48
836479     6039      2324      5    450
836480     6039       318      5   3628
836481     6039      1721      5     27
836482     6039      2571      5    124

[836483 rows x 4 columns]

In [426]:
from scipy.sparse import csr_matrix

# CSR Matrix 생성
num_user = ratings['user_id'].nunique()
num_movie = ratings['title'].nunique()

csr_data = csr_matrix((ratings['count'], (ratings.user_id, ratings.title)), shape= (num_user, num_movie))
csr_data

<6040x3629 sparse matrix of type '<class 'numpy.longlong'>'
	with 836483 stored elements in Compressed Sparse Row format>

In [427]:
from implicit.als import AlternatingLeastSquares
import os
import numpy as np

# implicit 라이브러리에서 권장하고 있는 부분입니다. 학습 내용과는 무관합니다.
os.environ['OPENBLAS_NUM_THREADS']='1'
os.environ['KMP_DUPLICATE_LIB_OK']='True'
os.environ['MKL_NUM_THREADS']='1'

In [428]:
# Implicit AlternatingLeastSquares 모델의 선언
als_model = AlternatingLeastSquares(factors=100, regularization=0.01, use_gpu=False, iterations=15, dtype=np.float32)

In [429]:
# als 모델은 input으로 item X user 꼴의 matrix를 받기 때문에 Transpose해줍니다.
csr_data_transpose = csr_data.T
csr_data_transpose

<3629x6040 sparse matrix of type '<class 'numpy.longlong'>'
	with 836483 stored elements in Compressed Sparse Column format>

4) CSR matrix를 직접 만들어 봅시다.

In [430]:
# 모델 훈련
als_model.fit(csr_data_transpose)

  0%|          | 0/15 [00:00<?, ?it/s]

In [432]:

# 훈련된 모델이 만든 나의 벡터와 영화 벡터 구하기
my_vector, matrix = user_to_idx[6041], movie_to_idx['Matrix, The (1999)']
my_vector, matrix_vector = als_model.user_factors[my_vector], als_model.item_factors[matrix]

In [433]:
my_vector

array([ 1.22833335e+00,  5.62239215e-02, -1.59503385e-01, -1.84889898e-01,
       -1.68979481e-01, -1.22353330e-01,  5.43722630e-01, -4.75693122e-02,
       -3.78764898e-01,  1.08256984e+00,  3.49917680e-01,  1.73650086e-01,
       -6.38982713e-01,  4.06543374e-01, -5.75075984e-01,  1.19217755e-02,
       -1.12571216e+00, -9.31851938e-02,  3.72354329e-01,  4.67013538e-01,
        1.51442021e-01,  5.52071571e-01,  3.53909194e-01,  7.51697600e-01,
       -3.21185254e-02,  6.82726979e-01, -1.98759526e-01, -2.10195243e-01,
        1.49604063e-02, -1.28654987e-01,  4.93119657e-01, -3.40344220e-01,
       -9.28556204e-01,  7.36328736e-02, -1.07435822e+00, -2.92284042e-01,
       -3.29814136e-01, -4.88126546e-01,  8.33079040e-01, -1.47266045e-01,
       -3.69759887e-01, -1.08713157e-01, -3.21160585e-01, -3.55445325e-01,
        1.37913799e+00, -1.71435475e-02,  5.18075377e-02, -4.99633908e-01,
        7.13892341e-01,  9.01375934e-02, -1.04771398e-01,  5.90528488e-01,
       -9.65893507e-01,  

In [434]:
matrix_vector

array([ 0.03692868,  0.00738154, -0.01479881,  0.01444819, -0.00449592,
       -0.00522446,  0.03562076,  0.01859645,  0.01086218,  0.01749213,
        0.01856558,  0.01773844, -0.02993364,  0.01058382,  0.01998881,
       -0.00282939, -0.02316605,  0.01237214, -0.00851961,  0.01390832,
       -0.0001294 ,  0.04753195,  0.01771171, -0.00427343,  0.00971743,
        0.02324702, -0.01419245,  0.00578098,  0.0015161 , -0.0187404 ,
        0.02211158, -0.01960597, -0.01846311,  0.00287443, -0.0032173 ,
        0.00583138,  0.00193565,  0.00693994,  0.03309618,  0.0147235 ,
       -0.00639107,  0.00447477,  0.00699209,  0.00953454,  0.02356692,
        0.01146753,  0.0071033 , -0.01949338,  0.02739416, -0.00084121,
        0.01860807,  0.020066  , -0.00734027, -0.01230309,  0.01718184,
        0.01605835,  0.01713888,  0.02898002,  0.00323494, -0.00989487,
        0.02307235, -0.00056675, -0.00905778,  0.00780105,  0.02943735,
       -0.01083852,  0.013817  , -0.0068008 ,  0.01643547,  0.00

In [438]:
movie_to_idx

{"One Flew Over the Cuckoo's Nest (1975)": 0,
 'James and the Giant Peach (1996)': 1,
 'My Fair Lady (1964)': 2,
 'Erin Brockovich (2000)': 3,
 "Bug's Life, A (1998)": 4,
 'Princess Bride, The (1987)': 5,
 'Ben-Hur (1959)': 6,
 'Christmas Story, A (1983)': 7,
 'Snow White and the Seven Dwarfs (1937)': 8,
 'Wizard of Oz, The (1939)': 9,
 'Beauty and the Beast (1991)': 10,
 'Gigi (1958)': 11,
 'Miracle on 34th Street (1947)': 12,
 "Ferris Bueller's Day Off (1986)": 13,
 'Sound of Music, The (1965)': 14,
 'Airplane! (1980)': 15,
 'Tarzan (1999)': 16,
 'Bambi (1942)': 17,
 'Awakenings (1990)': 18,
 'Big (1988)': 19,
 'Pleasantville (1998)': 20,
 'Wallace & Gromit: The Best of Aardman Animation (1996)': 21,
 'Back to the Future (1985)': 22,
 "Schindler's List (1993)": 23,
 'Meet Joe Black (1998)': 24,
 'Pocahontas (1995)': 25,
 'E.T. the Extra-Terrestrial (1982)': 26,
 'Titanic (1997)': 27,
 'Ponette (1996)': 28,
 'Close Shave, A (1995)': 29,
 'Antz (1998)': 30,
 'Girl, Interrupted (1999)':

In [439]:
favorite_movie = 'Saving Private Ryan (1998)'
movie_id = movie_to_idx[favorite_movie]
similar_movie = als_model.similar_items(movie_id, N=15)
similar_movie   # (movie idx, 유사도) Tuple

[(48, 1.0),
 (87, 0.66596717),
 (23, 0.651479),
 (157, 0.50511384),
 (3628, 0.46879748),
 (487, 0.44393817),
 (248, 0.43427),
 (141, 0.4234534),
 (3499, 0.40986493),
 (121, 0.39208636),
 (988, 0.38359687),
 (99, 0.36282024),
 (3493, 0.34780887),
 (75, 0.3449887),
 (124, 0.34384015)]

In [440]:
# movie_to_idx 를 뒤집어, index로부터 movie 이름을 얻는 dict를 생성합니다. 
idx_to_movie = {v:k for k,v in movie_to_idx.items()}
[idx_to_movie[i[0]] for i in similar_movie]

['Saving Private Ryan (1998)',
 'Braveheart (1995)',
 "Schindler's List (1993)",
 'Shawshank Redemption, The (1994)',
 'The Shawshank Redemption (1994)',
 'Boat, The (Das Boot) (1981)',
 'Good Will Hunting (1997)',
 'Fugitive, The (1993)',
 'Simon Sez (1999)',
 'Silence of the Lambs, The (1991)',
 'Glory (1989)',
 'American Beauty (1999)',
 'Held Up (2000)',
 'Hunt for Red October, The (1990)',
 'Matrix, The (1999)']

In [441]:
# 비슷한 영화를 알려주는 함수
def get_similar_movie(movie_title: str):
    movie_id = movie_to_idx[movie_title]
    similar_movie = als_model.similar_items(movie_id)
    similar_movie = [idx_to_movie[i[0]] for i in similar_movie]
    return similar_movie

In [443]:
get_similar_movie('Mission: Impossible (1996)')

['Mission: Impossible (1996)',
 'True Lies (1994)',
 'GoldenEye (1995)',
 'Lost World: Jurassic Park, The (1997)',
 'Twister (1996)',
 'Conspiracy Theory (1997)',
 'Con Air (1997)',
 'Young Sherlock Holmes (1985)',
 'Mask of Zorro, The (1998)',
 'Rock, The (1996)']

In [444]:
get_similar_movie('2001: A Space Odyssey (1968)')

['2001: A Space Odyssey (1968)',
 'Close Encounters of the Third Kind (1977)',
 'Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb (1963)',
 'Clockwork Orange, A (1971)',
 'Alien (1979)',
 '2010 (1984)',
 'Brazil (1985)',
 'Blade Runner (1982)',
 'Rear Window (1954)',
 'Day the Earth Stood Still, The (1951)']

In [445]:
user = user_to_idx[6041]
# recommend에서는 user*item CSR Matrix를 받습니다.
movie_recommended = als_model.recommend(user, csr_data, N=20, filter_already_liked_items=True)
movie_recommended

[(23, 0.44427347),
 (87, 0.39158636),
 (92, 0.336244),
 (60, 0.31596318),
 (248, 0.29368407),
 (146, 0.24877766),
 (107, 0.24852832),
 (64, 0.24412411),
 (99, 0.24071482),
 (141, 0.23984663),
 (44, 0.22345932),
 (160, 0.21519914),
 (157, 0.20712374),
 (988, 0.20661072),
 (375, 0.20652787),
 (175, 0.19950634),
 (154, 0.19587344),
 (38, 0.18915038),
 (145, 0.18702066),
 (126, 0.18557128)]

In [446]:
# 인덱스를 영화 제목으로 변환
[idx_to_movie[i[0]] for i in movie_recommended]

["Schindler's List (1993)",
 'Braveheart (1995)',
 'Terminator 2: Judgment Day (1991)',
 'Star Wars: Episode I - The Phantom Menace (1999)',
 'Good Will Hunting (1997)',
 'Thelma & Louise (1991)',
 'Jurassic Park (1993)',
 'Star Wars: Episode VI - Return of the Jedi (1983)',
 'American Beauty (1999)',
 'Fugitive, The (1993)',
 'Star Wars: Episode IV - A New Hope (1977)',
 'Forrest Gump (1994)',
 'Shawshank Redemption, The (1994)',
 'Glory (1989)',
 'Face/Off (1997)',
 'Men in Black (1997)',
 'As Good As It Gets (1997)',
 'Sixth Sense, The (1999)',
 'Fifth Element, The (1997)',
 'Shakespeare in Love (1998)']

In [448]:
# 사용자가 기록을 남긴 데이터가 특정 영화 추천에 기여한 정도를 확인
recommended = movie_to_idx['Star Wars: Episode IV - A New Hope (1977)']
explain = als_model.explain(user, csr_data, itemid=recommended)

[(idx_to_movie[i[0]], i[1]) for i in explain[1]]

[('Matrix, The (1999)', 0.10651521843673317),
 ('Saving Private Ryan (1998)', 0.07141647414309976),
 ('Titanic (1997)', 0.04775964962356534),
 ('The Shawshank Redemption (1994)', -2.1723490017454678e-05),
 ('Life Is Beautiful (La Vita è bella) (1997)', -0.008512690267092025)]